### ISSUE TITLE
NuGet package with DLL

### ISSUE THREAD

**Issue by pmg23** [stakx/TypeNameFormatter#22]
We've adopted TypeNameFormatter now in our codebase (thanks!) - on version 1.0.0-beta for now but looking forward to the next version.

Unfortunately the road has been less than smooth because of the apparently poor support for source-only NuGet packages in both Visual Studio/ReSharper and Rider (as of the latest versions).

For Rider, see https://youtrack.jetbrains.com/issue/RIDER-16638.

For Visual Studio/ReSharper, I gather the problem is that [ReSharper Build](https://blog.jetbrains.com/dotnet/2015/10/15/introducing-resharper-build/) doesn't include the packaged sources correctly, if at all, causing the build to fail.

MSBuild seems to work perfectly.

I wonder please:
- have you successfully integrated TypeNameFormatter.Sources (or other similarly-packaged libraries) into existing solutions?
- would you consider providing a vanilla NuGet package containing a DLL targeting AnyCPU and (say) net47?

Regards
Matt

**Comment by stakx** [comment]
Hey, great to hear this small lib is useful to you! :heart:

> have you successfully integrated TypeNameFormatter.Sources (or other similarly-packaged libraries) into existing solutions?

As of now, not yet. You'd be the first. It might end up in Moq v5, but I'm still waiting on feedback there.

Anyways your feedback is exactly what I'm looking for and it wpuld be nice to get this working with toolchains other than plain MSBuild.

> would you consider providing a vanilla NuGet package containing a DLL targeting AnyCPU and (say) net47?

Sure, I just haven't got around to setting up the solution for two different packages yet, but I'm definitely planning to add a binary package targeting pretty much every target framework after `net20`. That's why the current NuGet has the weird `.Sources` suffix... the binsry package will be named `TypeNameFormatter`. I might get around to it later this week.

**Comment by stakx** [comment]
Regarding the source only distribution, perhaps it would be better to choose a more straight-forward approach and just put the source files in the NuGet package's `content\` folder? I'm not sure if there's a future for that, I've read hints that this might not be supported much longer, and it has the side effect of visibly adding files to your projects... but tooling might understand this form of distributiom better than hooking into the `CoreCompile` build target.

**Comment by stakx** [comment]
Just a brief update: I've been playing around with Rider a little, and I think I might be able to get things working better by rewriting the NuGet packages `.targets` file. Will fiddle some more tomorrow. Stay tuned!

**Comment by stakx** [comment]
@pmg23, I just published [`1.0.0-beta2`](https://www.nuget.org/packages/TypeNameFormatter.Sources/1.0.0-beta2). It's got all the changes we've discussed, and hopefully it'll work more smoothly with Rider and Visual Studio + ReSharper. I haven't yet been able to test the latter combination, as I don't have ReSharper installed.

As always, I'm looking forward to your feedback.

**Comment by pmg23** [comment]
Hi. Thanks so much for working on this. I tried out TypeNameFormatter.Sources 1.0.0-beta2 from NuGet and have good and bad news.

The good is that it appears to work perfectly in both Rider and Visual Studio with ReSharper.

The bad is that it breaks MSBuild *for downstream projects*. Specifically, I have project A with (simplified)
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <AssemblyName>A</AssemblyName>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Castle.Windsor" Version="4.1.0" />
    <PackageReference Include="TypeNameFormatter.Sources" Version="1.0.0-beta2" />
  </ItemGroup>
</Project>
```
and B with
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <AssemblyName>B</AssemblyName>
  </PropertyGroup>
  <ItemGroup>
    <ProjectReference Include="..\A\A.csproj" />
  </ItemGroup>
</Project>
```

Note that B automagically inherits A's transitive package dependencies, so e.g. it can use symbols from Windsor without declaring an explicit dependency.

Unfortunately now MSBuild fails on compiling B with
```
CSC : error CS0006: Metadata file '...\B\obj\TypeNameFormatter.cs' could not be found [...\B\B.csproj]
```

If I understand it right, B has inherited the source file dependency from A (which is in itself undesirable though harmless) but does not have its own copy of the file.

Best regards
Matt

**Comment by stakx** [comment]
@pmg23, first of all, thanks once again for your valuable feedback. Good to hear that at least Rider and ReSharper are now happy.

MSBuild is always a bit of a (bad) adventure, but I think I've identified the problem with the downstream build breakage. Sorry for that fuck-up. Expect a fixed `1.0.0-beta3` soon.

**Comment by stakx** [comment]
@pmg23 - here we go again: [`1.0.0-beta3`](https://www.nuget.org/packages/TypeNameFormatter.Sources/1.0.0-beta3). I'm crossing my fingers at this point. 😄

**Comment by pmg23** [comment]
beta3 looks good! (MSBuild issue fixed, and still fine in Rider and VS/R#.) Thanks again.

**Comment by stakx** [comment]
Wheeeee! Really glad to hear. Thanks for letting me know.

**Comment by stakx** [comment]
(Just as a small update, I am currently blocked from publishing a binary NuGet package to NuGet.org. Because I've already uploaded a package called `TypeNameFormatter.Sources` it appears I can no longer upload `TypeNameFormatter`... which is a bit vexing. Still in contact with NuGet support to resolve this.)

**Comment by pmg23** [comment]
Hi. Belated thanks for the update. I wonder if you have had any luck with NuGet support?

(To be clear, as the source package integration works fine and suits our use case, I am planning to use that rather than the binary package. I'm only blocked pending an "official" (non-prerelease) version.)

**Comment by stakx** [comment]
@pmg23: I haven't heard back from NuGet, but I believe they'd have to physically delete the packages I've already uploaded, which I'd hate to do. So I'll probably keep this source-only.

I'll release 1.0.0 proper during the next 48 hours or so.

**Comment by stakx** [comment]
@pmg23 - just published [`1.0.0` on NuGet.org](https://www.nuget.org/packages/TypeNameFormatter.Sources/1.0.0), it should become available shortly once it's indexed.

I'm closing this issue, since I probably won't do a binary release anytime soon (due to the NuGet blockage mentioned above).

**Comment by pmg23** [comment]
Using 1.0.0 in our application now. Thanks!
